[SPARK-43487][SQL] Fix Nested CTE error message

johanl-db opened the pull request:
### What changes were proposed in this pull request?
The batch of errors migrated to error classes as part of spark-40540 contains an error that got mixed up with the wrong error message:

[ambiguousRelationAliasNameInNestedCTEError](https://github.com/apache/spark/commit/43a6b932759865c45ccf36f3e9cf6898c1b762da#diff-744ac13f6fe074fddeab09b407404bffa2386f54abc83c501e6e1fe618f6db56R1983) uses the same error message as the following commandUnsupportedInV2TableError:

WITH t AS (SELECT 1), t2 AS ( WITH t AS (SELECT 2) SELECT * FROM t) SELECT * FROM t2;
AnalysisException: t is not supported for v2 tables

The error should be:

AnalysisException: Name tis ambiguous in nested CTE.
Please set spark.sql.legacy.ctePrecedencePolicy to CORRECTED so that name defined in inner CTE takes precedence. If set it to LEGACY, outer CTE definitions will take precedence. See more details in SPARK-28228.

### Does this PR introduce _any_ user-facing change?
Fix user-facing error message for ambiguous name in nested CTEs.


### How was this patch tested?

johanl-db committed:
Fix Nested CTE error message

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3783,6 +3783,12 @@
       "Failed to execute command because subquery expressions are not allowed in DEFAULT values."
     ]
   },
+  "_LEGACY_ERROR_TEMP_1348" : {
```
Could you assign proper name to the error class, please. The `_LEGACY_ERROR_TEMP_` prefix was used for migration purpose in batch way to switch to error classes in all existing exceptions.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3233,6 +3225,14 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+    new AnalysisException(
+      errorClass = "_LEGACY_ERROR_TEMP_1348",
+      messageParameters = Map(
+        "name" -> name,
```
Please, wrap it by `toSQLId`, please.

johanl-db replied:
👍

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3233,6 +3225,14 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+      errorClass = "_LEGACY_ERROR_TEMP_1348",
+      messageParameters = Map(
+        "name" -> name,
+        "config" -> LEGACY_CTE_PRECEDENCE_POLICY.key))
```
Use `toSQLConf`

johanl-db replied:
👍

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -3783,6 +3783,12 @@
+  "_LEGACY_ERROR_TEMP_1348" : {
+    "message" : [
+      "Name <name> is ambiguous in nested CTE.",
+      "Please set <config> to CORRECTED so that name defined in inner CTE takes precedence. If set it to LEGACY, outer CTE definitions will take precedence. See more details in SPARK-28228."
```
```suggestion
      "Please set <config> to \"CORRECTED\" so that name defined in inner CTE takes precedence. If set it to \"LEGACY\", outer CTE definitions will take precedence. See more details in SPARK-28228."
```

MaxGekk replied:
> See more details in SPARK-28228.

It would be better to update the public docs and refer to them.

johanl-db replied:
I linked to the existing documentation that covers this configuration in the migration guide: https://spark.apache.org/docs/latest/sql-migration-guide.html#query-engine

MaxGekk commented:
@johanl-db Are you working on the PR? Could you, please, address the comments above.

johanl-db committed:
Add link to doc & test

johanl-db commented:
> @johanl-db Are you working on the PR? Could you, please, address the comments above.

I was away the past few days, I'm picking this up now. I addressed your comments, please take another look.

MaxGekk commented:
+1, LGTM. Merging to master.
Thank you, @johanl-db.
